internal/statestore/backfill_test.go

```diff
@@ -301,3 +301,122 @@ func TestDeleteBackfill(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+
+	t.Run("WithValidContext", func(t *testing.T) {
+		ctx := utilTesting.NewContext(t)
+		generateBackfills(ctx, t, service, 2)
+		c, err := redis.Dial("tcp", fmt.Sprintf("%s:%s", cfg.GetString("redis.hostname"), cfg.GetString("redis.port")))
+		require.NoError(t, err)
+		idsIndexed, err := redis.Strings(c.Do("HKEYS", allBackfills))
+		require.NoError(t, err)
+		require.Len(t, idsIndexed, 2)
+		require.Equal(t, "mockBackfillID-0", idsIndexed[0])
+		require.Equal(t, "mockBackfillID-1", idsIndexed[1])
+	})
+
+	t.Run("WithCancelledContext", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+		service = New(cfg)
+		err := service.IndexBackfill(ctx, &pb.Backfill{
+			Id:         "12345",
+			Generation: 42,
+		})
+		require.Error(t, err)
+		require.Equal(t, codes.Unavailable.String(), status.Convert(err).Code().String())
+		require.Contains(t, status.Convert(err).Message(), "IndexBackfill, id: 12345, failed to connect to redis:")
+	})
+}
+
+func TestDeindexBackfill(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+
+	ctx := utilTesting.NewContext(t)
+
+	generateBackfills(ctx, t, service, 2)
+
+	c, err := redis.Dial("tcp", fmt.Sprintf("%s:%s", cfg.GetString("redis.hostname"), cfg.GetString("redis.port")))
+	require.NoError(t, err)
+	idsIndexed, err := redis.Strings(c.Do("HKEYS", allBackfills))
+	require.NoError(t, err)
+	require.Len(t, idsIndexed, 2)
+	require.Equal(t, "mockBackfillID-0", idsIndexed[0])
+	require.Equal(t, "mockBackfillID-1", idsIndexed[1])
+
+	// deindex and check that there is only 1 backfill in the returned slice
+	err = service.DeindexBackfill(ctx, "mockBackfillID-1")
+	require.NoError(t, err)
+	idsIndexed, err = redis.Strings(c.Do("HKEYS", allBackfills))
+	require.NoError(t, err)
+	require.Len(t, idsIndexed, 1)
+	require.Equal(t, "mockBackfillID-0", idsIndexed[0])
+
+	// pass an expired context, err expected
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	service = New(cfg)
+	err = service.DeindexBackfill(ctx, "12345")
+	require.Error(t, err)
+	require.Equal(t, codes.Unavailable.String(), status.Convert(err).Code().String())
+	require.Contains(t, status.Convert(err).Message(), "DeindexBackfill, id: 12345, failed to connect to redis:")
+}
+
+func TestGetIndexedBackfills(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+
+	ctx := utilTesting.NewContext(t)
+
+	verifyBackfills := func(service Service, backfills []*pb.Backfill) {
+		ids, err := service.GetIndexedBackfills(ctx)
+		require.Nil(t, err)
+		require.Equal(t, len(backfills), len(ids))
+
+		for _, tt := range backfills {
+			_, ok := ids[tt.GetId()]
```

Comment:
Can you please add one more test case when you Index a new Backfill with Generation > 1 and get an output
```suggestion
			gen, ok := ids[tt.GetId()]
			require.Equal(t, tt.Generation, gen)
```

Reply:
Applied changes